torchrl/collectors/distributed/distributed_collector.py

```diff
@@ -0,0 +1,232 @@
+
+        self.collected_frames = 0
+        self.total_frames = total_frames
+        self.collector_class = collector_class
+        self.collector_params = collector_params
+        self.num_collectors = num_collectors
+        self.remote_config = remote_config
+        self.communication = communication
+
+        # Create a local instance of the collector class.
+        # Will be used to track latest policy weights.
+        self._local_collector = self._make_collector(
+            self.collector_class, collector_params)
+
+        # Create remote instances of the collector class
+        self._remote_collectors = []
+        if self.num_collectors > 0:
+            self.add_collectors(self.num_collectors, collector_params)
+
+        # Print info of all remote workers
+        pending_samples = [e.print_remote_collector_info.remote() for e in self.remote_collectors()]
+        ray.wait(pending_samples)
+
+    @staticmethod
+    def _make_collector(cls, collector_params):
+        """Create a single collector instance."""
+        w = cls(**collector_params)
+        return w
+
+    def add_collectors(self, num_collectors, collector_params):
+        """Create and add a number of remote collectors to the set."""
+        cls = self.collector_class.as_remote(**self.remote_config).remote
+        self._remote_collectors.extend(
+            [self._make_collector(cls, collector_params) for _ in range(num_collectors)])
+
+    def local_collector(self):
+        """Return local collector"""
+        return self._local_collector
+
+    def remote_collectors(self):
+        """Returns list of remote collectors"""
+        return self._remote_collectors
+
+    def stop(self):
+        """Stop all remote collectors"""
+        for w in self.remote_collectors():
+            w.__ray_terminate__.remote()
+
+    def __iter__(self) -> Iterator[TensorDictBase]:
+        if self.communication == "sync":
+            return self.sync_iterator()
+        else:
+            return self.async_iterator()
+
+    def sync_iterator(self) -> Iterator[TensorDictBase]:
+
+        while self.collected_frames < self.total_frames:
+
+            # Broadcast agent weights
+            self.local_collector().update_policy_weights_()
+            state_dict = {"policy_state_dict": self._local_collector.policy.state_dict()}
+            state_dict = ray.put(state_dict)
+            for e in self.remote_collectors():
+                e.load_state_dict.remote(**{"state_dict": state_dict, "strict": False})
+
+            # Ask for batches to all remote workers.
+            pending_samples = [e._iterator_step.remote() for e in self.remote_collectors()]
+
+            # Wait for all rollouts
+            samples_ready = []
+            while len(samples_ready) < self.num_collectors:
+                samples_ready, samples_not_ready = ray.wait(
+                    pending_samples, num_returns=len(pending_samples), timeout=0.001)
```

Comment:
what's the role of timeout here? Is the value standard? Should it be a hyperparam or a global variable?

Reply:
If timeout is set, the function returns either when the requested number of IDs are ready or when the timeout is reached, whichever occurs first. If it is not set, the function simply waits until that number of objects is ready and returns that exact number of object refs.

Actually for not we can just set timeout=None, but if we want to implement the  preemption mechanism, we will need to check regularly how many of the tasks have ended. We can either set it to a reasonable value of allow the users to define it.